drivers/dma/dma_intel_adsp_hda.c

```diff
@@ -233,6 +233,15 @@ int intel_adsp_hda_dma_status(const struct device *dev, uint32_t channel,
 	stat->pending_length = used;
 	stat->free = unused;
 
+#if CONFIG_SOC_INTEL_ACE20_LNL || CONFIG_SOC_INTEL_ACE30_PTL
+	if (cfg->direction == MEMORY_TO_PERIPHERAL || cfg->direction == PERIPHERAL_TO_MEMORY) {
+		uint32_t llp_l, llp_u;
+
+		llp_l = *DGLLLPL(cfg->base, cfg->regblock_size, channel);
+		llp_u = *DGLLLPU(cfg->base, cfg->regblock_size, channel);
+		stat->total_copied = ((uint64_t)llp_u << 32) | llp_l;
```

Comment:
isn't there a potential problem with this code?

if llp_l is MAX_32, then llp_u will likely be increased by the time you read it, but the llp_l be close to zero. This sort of combination carries the risk of creating completely wrong counts. I think you need to re-read the llp_l and make sure its value did not become lower than before...

Reply:
@plbossart, it is copied directly from the existing `intel_adsp_gpdma_llp_read()` from
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/dma/dma_intel_adsp_gpdma.c#L133

I see, I will update, thanks for spotting it.

Reply:
wow, I earned by salary for the day. GPDMA is also wrong haha.

Reply:
probably need something along those lines

````diff
diff --git a/drivers/dma/dma_intel_adsp_gpdma.c b/drivers/dma/dma_intel_adsp_gpdma.c
index 5e3850244d9b..a3e24e994ad5 100644
--- a/drivers/dma/dma_intel_adsp_gpdma.c
+++ b/drivers/dma/dma_intel_adsp_gpdma.c
@@ -129,9 +129,15 @@ static inline void intel_adsp_gpdma_llp_read(const struct device *dev,
 {
 #ifdef CONFIG_DMA_INTEL_ADSP_GPDMA_HAS_LLP
        const struct intel_adsp_gpdma_cfg *const dev_cfg = dev->config;
+       uint32_t llp_l1;
+       uint32_t llp_l2;
 
-       *llp_l = dw_read(dev_cfg->shim, GPDMA_CHLLPL(channel));
+       llp_l1 = dw_read(dev_cfg->shim, GPDMA_CHLLPL(channel));
        *llp_u = dw_read(dev_cfg->shim, GPDMA_CHLLPU(channel));
+       llp_l2 = dw_read(dev_cfg->shim, GPDMA_CHLLPL(channel));
+       if (llp_l2 < llp_l1)
+               *llp_u = dw_read(dev_cfg->shim, GPDMA_CHLLPU(channel));
+       *llp_l = llp_l2;
 #endif
 }


````

Reply:
Ah yes... the ole double word counter read, good catch!

Reply:
@ujfalusi do you want to send the same fix for GPDMA? same problem.